### PR TITLE
Split _cache_key

### PR DESCRIPTION
**Pull request by jonathanchu33**

### CODE REVIEW

**Comment by jonathanchu33** [comment]
Benchmarks over 3 batches of 10 trials each

Minimum
defun_matmul_2_by_2_CPU: 144 -> 147 us
defun_matmul_2_by_2_CPU_async: 80.5 -> 80.5 us
defun_matmul_2_by_2_with_signature_CPU: 198 -> 194 us

Mean
defun_matmul_2_by_2_CPU: 163.5 -> 162 us
defun_matmul_2_by_2_CPU_async: 85 -> 85 us
defun_matmul_2_by_2_with_signature_CPU: 207.5 -> 214 us

Median
defun_matmul_2_by_2_CPU: 163.5-> 163 us
defun_matmul_2_by_2_CPU_async: 82.5 -> 86 us
defun_matmul_2_by_2_with_signature_CPU: 206 -> 201 us

Inconclusive

**Review by sun51 — COMMENTED** [review]
In Line 3307, we should split that to 2 parts. so that the cache_key_context is saved, and then passed down to _define_function_with_shape_relaxation in line 3344. Then inside line 3344, we do not have to calculate the cache_key_context again

**Review by sun51 — COMMENTED** [review]
And this optimization should improve benchmark_defun_matmul_2_by_2_relaxed_shape_CPU benchmark, but for other benchmarks, it should not affect much.

**Comment by jonathanchu33** [comment]
Updated:
Benchmarks over 3 batches of 10 trials each

Minimum
defun_matmul_2_by_2_CPU: 144 -> 145.5 us
defun_matmul_2_by_2_CPU_async: 80.5 -> 80 us
defun_matmul_2_by_2_with_signature_CPU: 198 -> 187 us

Mean
defun_matmul_2_by_2_CPU: 163.5 -> 153 us
defun_matmul_2_by_2_CPU_async: 85 -> 82.5 us
defun_matmul_2_by_2_with_signature_CPU: 207.5 -> 195 us

Median
defun_matmul_2_by_2_CPU: 163.5-> 151 us
defun_matmul_2_by_2_CPU_async: 82.5 -> 82.5 us
defun_matmul_2_by_2_with_signature_CPU: 206 -> 194 us

Apparent improvement

File: tensorflow/python/eager/function.py [hunk 1]
```diff
@@ -3073,7 +3073,8 @@ def __get__(self, instance, owner):
     # Return the cached `Function` for the instance
     return self._descriptor_cache[instance]
 
-  def _cache_key(self, args, kwargs, include_tensor_ranks_only=False):
+  def _cache_key(self, args, kwargs, include_tensor_ranks_only=False,
+                 cache_key_context=None):
```
[the remark this answers is not in the document]
Comment (jonathanchu33) [reply tensorflow/python/eager/function.py]: Done

**Review by sun51 — DISMISSED** [review]

**Review by rthadur — DISMISSED** [review]

**Review by kkimdev — APPROVED** [review]

**Review by sun51 — APPROVED** [review]

**Pull request merged**